toolchain/check/handle_where.cpp

```diff
@@ -8,41 +8,78 @@
 
 namespace Carbon::Check {
 
-auto HandleParseNode(Context& context, Parse::SelfTypeNameId node_id) -> bool {
-  return context.TODO(node_id, "HandleSelfTypeName");
-}
-
-auto HandleParseNode(Context& context, Parse::DesignatorExprId node_id)
+auto HandleParseNode(Context& context, Parse::WhereOperandId /*node_id*/)
     -> bool {
-  return context.TODO(node_id, "HandleDesignatorExpr");
-}
+  // The expression at the top of the stack represents a constraint type that
+  // is being modified by the `where` operator. It would be `MyInterface` in
+  // `MyInterface where .Member = i32`.
+  auto [self_node, self_id] = context.node_stack().PopExprWithNodeId();
+  auto self_type_id = ExprAsType(context, self_node, self_id);
+  // TODO: Do this instead once `WhereExpr` is ready to consume this:
+  // context.node_stack().Push(node_id, self_type_id);
+  context.node_stack().Push(self_node, self_id);
 
-auto HandleParseNode(Context& context, Parse::WhereOperandId node_id) -> bool {
-  return context.TODO(node_id, "HandleWhereOperand");
+  // Introduce a name scope so that we can remove the `.Self` entry we are
+  // adding to name lookup at the end of the `where` expression.
+  context.scope_stack().Push();
+  // Introduce `.Self` as a symbolic binding. Its type is the value of the
+  // expression to the left of `where`, so `MyInterface` in the example above.
+  // Because there is no equivalent non-symbolic value, we use `Invalid` as
+  // the `value_id` on the `BindSymbolicName`.
+  auto entity_name_id = context.entity_names().Add(
+      {.name_id = SemIR::NameId::PeriodSelf,
+       .parent_scope_id = context.decl_name_stack().PeekParentScopeId(),
+       .bind_index = context.scope_stack().AddCompileTimeBinding()});
+  // FIXME: should this have a location associated with `node_id`?
```

Comment:
`node_id` has the location of the `where` keyword (with the type to the left as its child), which is where the `.Self` name was introduced. `self_node` is the type to the left of `where`, which is relevant but the connection is less direct.

Reply:
`node_id` might have a more direction connection because it informs the value of `.Self` for the RHS, but how do you anticipate the location being used for diagnostics or debug line numbers?

For example, in my head if we have a declaration like:

```
impl forall [U:! type, T:! AddWith(U) where .Self impls AssignWith(.Self.Result)]
     T as AddAssignWith(U) {
```

I could see generating a diagnostic like:

```
error: `Result` is not a member of `.Self`
impl forall [U:! type, T:! AddWith(U) where .Self impls AssignWith(.Self.Result)]
                                                                         ^~~~~~
note: `.Self` defined as `AddWith(U)` here
impl forall [U:! type, T:! AddWith(U) where .Self impls AssignWith(.Self.Result)]
                           ^~~~~~~~~~
```

IIUC, that would be using `self_node` to directly get `AddWith(U)`? Can you give an example of how you'd expect `node_id` to be used if it were in the location?

Reply:
I should note too, maybe we'd just say "`Result` is not a member of `AddWith(U)`" and never reference the origin of that type. That's a possibility too, and I think part of why we end up not using `Self` locations in most cases.

Reply:
My thought was that the error would be like what you wrote, but also include `where`.

```
note: `.Self` defined as `AddWith(U)` here
impl forall [U:! type, T:! AddWith(U) where .Self impls AssignWith(.Self.Result)]
                           ^~~~~~~~~~~~~~~~
```

I think any of the three options is reasonable though.

Reply:
Thinking about this further, I think there is another example we should consider:
```
fn F[T:! AddWith(.Self)](x: T, y: T)
```
In this case the `.Self` is introduced by the `:!`. Furthermore, the current idea is to mark the `T` node as being part of the compile-time binding since the `:!` is output at the end of the expression (unless we were to make it a virtual node too). So I think we should go with the no-node approach for now, since that is all I can easily support with this other case.